python/ray/data/tests/test_dynamic_block_split.py

```diff
@@ -3,67 +3,268 @@
+
+    try:
         ds = ray.data.read_datasource(
-            LargeBytesDatasource(),
-            parallelism=1,
+            RandomBytesDatasource(),
+            parallelism=num_tasks,
+            num_batches=num_batches,
+            batch_size=batch_size,
         )
+        dsp = ds.window(blocks_per_window=2)
+        assert dsp._length == num_tasks / 2
+
+        dsp = dsp.map_batches(lambda x: x)
+        result_batches = list(ds.iter_batches(batch_size=5))
+        for batch in result_batches:
+            assert len(batch) == 5
+        assert len(result_batches) == num_batches * num_tasks / 5
 
-        ds = ds.map_batches(foo, batch_size=None)
-        assert ds.count() == num_batch
+        dsp = ds.window(blocks_per_window=2)
+        assert dsp._length == num_tasks / 2
+
+        dsp = ds.repeat().map_batches(lambda x: x)
+        assert len(dsp.take(5)) == 5
     finally:
-        ctx.block_splitting_enabled = block_splitting_enabled
+        ctx.target_max_block_size = target_max_block_size
+
+
+def test_lazy_block_list(shutdown_only, enable_dynamic_block_splitting):
+    # Test 10 blocks from 10 tasks, each block is 1024 bytes.
+    num_batches = 10
+    batch_size = 1024
+    num_tasks = 10
+
+    ctx = ray.data.context.DatasetContext.get_current()
+    target_max_block_size = ctx.target_max_block_size
+    ctx.target_max_block_size = batch_size
+
+    try:
+        ds = ray.data.read_datasource(
+            RandomBytesDatasource(),
+            parallelism=num_tasks,
+            num_batches=num_batches,
+            batch_size=batch_size,
+        )
+
+        # Check internal states of LazyBlockList before execution
+        block_list = ds._plan._in_blocks
+        block_refs = block_list._block_partition_refs
+        cached_metadata = block_list._cached_metadata
+        metadata = block_list.get_metadata()
+
+        assert isinstance(block_list, LazyBlockList)
+        assert len(block_refs) == num_tasks
+        assert block_refs[0] is not None and all(
+            map(lambda ref: ref is None, block_refs[1:])
+        )
+        assert all(map(lambda ref: ref is None, block_list._block_partition_meta_refs))
+        assert len(cached_metadata) == num_tasks
+        for i, block_metadata in enumerate(cached_metadata):
+            if i == 0:
+                assert len(block_metadata) == num_batches
+                for m in block_metadata:
+                    assert m.num_rows == 1
+            else:
+                assert block_metadata is None
+        assert len(metadata) == num_tasks - 1 + num_batches
+        for i, block_metadata in enumerate(metadata):
+            if i < num_batches:
+                assert block_metadata.num_rows == 1
+                assert block_metadata.schema is not None
+            else:
+                assert block_metadata.num_rows == num_batches
+                assert block_metadata.schema is None
+
+        # Check APIs of LazyBlockList
```

Comment:
It's nice to cover the APIs in unit test, but should they go to a separate file since these looks have little to do with block splitting?

Reply:
it's actually testing the APIs behavior when dynamic block splitting is enabled, so I feel it's okay to put in this file or another separate file (e.g. this file also tests API for Dataset and DatasetPipeline). We can refactor later to create a dedicated test file `test_lazy_block_list.py` when consolidating LazyBlockList and BlockList.